lib/http/client/client-base.js

```diff
@@ -0,0 +1,798 @@
+    this.read += length;
+
+    const { client, socket, body, read } = this;
+
+    const ret = body ? body(null, chunk.slice(offset, offset + length)) : null;
+
+    if (ret == null && read > client[kMaxAbortedPayload]) {
+      socket.destroy();
+    } else if (ret === false) {
+      socket.pause();
+    }
+  }
+
+  [HTTPParser.kOnMessageComplete]() {
+    const { body } = this;
+
+    this.read = 0;
+    this.body = null;
+
+    if (body) {
+      body(null, null);
+      this.next();
+    }
+  }
+
+  next() {
+    const { client, resumeSocket } = this;
+
+    resumeSocket();
+
+    client[kQueue][client[kRunningIdx]++] = null;
+
+    resume(client);
+  }
+
+  destroy(err) {
+    const { client, body } = this;
+
+    assert(err);
+
+    if (client[kRunningIdx] >= client[kPendingIdx]) {
+      assert(!body);
+      return;
+    }
+
+    this.read = 0;
+    this.body = null;
+
+    // Retry all idempotent requests except for the one
+    // at the head of the pipeline.
+
+    const retryRequests = [];
+    const errorRequests = [];
+
+    errorRequests.push(client[kQueue][client[kRunningIdx]++]);
+
+    const requests = client[kQueue].slice(client[kRunningIdx],
+                                          client[kPendingIdx]);
+    for (const request of requests) {
+      const { idempotent, body } = request;
+      if (idempotent && (!body || typeof body.pipe !== 'function')) {
+        retryRequests.push(request);
+      } else {
+        errorRequests.push(request);
+      }
+    }
+
+    client[kQueue].splice(0, client[kPendingIdx], ...retryRequests);
+
+    client[kPendingIdx] = 0;
+    client[kRunningIdx] = 0;
+
+    if (body) {
+      body(err, null);
+    }
+
+    for (const request of errorRequests) {
+      request.invoke(err, null);
+    }
+
+    this.close();
+  }
+}
+
+function connect(client) {
+  assert(!client[kSocket]);
+  assert(!client[kRetryTimeout]);
+
+  const { protocol, port, hostname } = client[kUrl];
+  const servername = client[kServerName] ||
+    (client[kTLSOpts] && client[kTLSOpts].servername);
+  const socket = protocol === 'https:' ?
+    tls.connect(port || 443, hostname, {
+      ...client[kTLSOpts],
+      servername
+    }) : net.connect(port || 80, hostname);
+
+  client[kSocket] = socket;
+
+  socket[kClient] = client;
+  socket[kParser] = new Parser(client, socket);
+  socket[kClosed] = false;
+  socket[kError] = null;
+  socket.setTimeout(client[kSocketTimeout], function() {
+    this.destroy(new Error('socket timeout'));
+  });
+  socket
+    .setNoDelay(true)
+    .on(protocol === 'https:' ? 'secureConnect' : 'connect', function() {
+      const client = this[kClient];
+
+      client[kRetryDelay] = 0;
+      client.emit('connect');
+      resume(client);
+    })
+    .on('data', function(chunk) {
+      const parser = this[kParser];
+
+      const err = parser.execute(chunk);
+      if (err instanceof Error && !this.destroyed) {
+        this.destroy(err);
+      }
+    })
```

Comment:
@addaleax it’s probably better if you review over at the undici repo. I wrote a comment about this above.

There has been little response here and a lot of progress since this PR was opened and I have not kept this up to date as it requires significant effort due to the difference in core style.

Reply:
FYI your comment has already been resolved 😄